File: api_fastapi/app/services/availability_service.py
from datetime import datetime, timedelta
from sqlalchemy.orm import Session
from app.models.reservation import Reservation, StatutReservation
from app.schemas.reservation import AvailabilityRequest, AvailabilityResponse
from app.config import settings


class AvailabilityService:
    """Service for checking table availability"""

    @staticmethod
    def check_availability(
        db: Session,
        request: AvailabilityRequest,
        exclude_reservation_id: int = None
    ) -> AvailabilityResponse:
        """
        Check if there's availability for the requested date and number of people
        Time window: ±1 hour from requested time
        """
        # Calculate time window (±1 hour)
        start_time = request.dateReservation - timedelta(hours=1)
        end_time = request.dateReservation + timedelta(hours=1)

        # Query reservations in the time window (excluding cancelled ones)
        query = db.query(Reservation).filter(
            Reservation.date_reservation >= start_time,
            Reservation.date_reservation <= end_time,
            Reservation.statut != StatutReservation.ANNULEE
        )

        # Exclude the current reservation if updating
        if exclude_reservation_id:
            query = query.filter(Reservation.id != exclude_reservation_id)

        reservations = query.all()

        # Calculate reserved seats
        reserved_seats = sum(r.nombre_personnes for r in reservations)

        # Calculate available seats
        total_capacity = settings.RESTAURANT_CAPACITY
        available_seats = total_capacity - reserved_seats

        # Check if requested number of people can be accommodated
        is_available = available_seats >= request.nombrePersonnes

        return AvailabilityResponse(
            dateReservation=request.dateReservation,
            totalCapacity=total_capacity,
            reservedSeats=reserved_seats,
            availableSeats=available_seats,
            available=is_available
        )

File: api/src/main/java/fr/esgi/api/verydarkkitchensmartphone/service/AvailabilityService.java
package fr.esgi.api.verydarkkitchensmartphone.service;

import fr.esgi.api.verydarkkitchensmartphone.dto.AvailabilityRequest;
import fr.esgi.api.verydarkkitchensmartphone.dto.AvailabilityResponse;
import fr.esgi.api.verydarkkitchensmartphone.models.Reservation;
import fr.esgi.api.verydarkkitchensmartphone.models.StatutReservation;
import fr.esgi.api.verydarkkitchensmartphone.repository.ReservationRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.util.List;

@Service
@RequiredArgsConstructor
public class AvailabilityService {

    private final ReservationRepository reservationRepository;

    @Value("${restaurant.capacity:50}")
    private Integer restaurantCapacity;

    /**
     * Check availability for a given time slot (2-hour window)
     * Considers reservations within ±1 hour of requested time
     */
    public AvailabilityResponse checkAvailability(AvailabilityRequest request) {
        LocalDateTime requestedTime = request.getDateReservation();
        Integer requestedSeats = request.getNombrePersonnes();

        // Define time slot window (1 hour before and after)
        LocalDateTime startWindow = requestedTime.minusHours(1);
        LocalDateTime endWindow = requestedTime.plusHours(1);

        // Get all active reservations in this time window
        List<Reservation> reservations = reservationRepository
                .findByDateReservationBetween(startWindow, endWindow);

        // Calculate total reserved seats (excluding cancelled reservations)
        int reservedSeats = reservations.stream()
                .filter(r -> r.getStatut() != StatutReservation.ANNULEE)
                .mapToInt(Reservation::getNombrePersonnes)
                .sum();

        int availableSeats = restaurantCapacity - reservedSeats;
        boolean isAvailable = availableSeats >= requestedSeats;

        return new AvailabilityResponse(
                requestedTime,
                restaurantCapacity,
                reservedSeats,
                availableSeats,
                isAvailable
        );
    }

    /**
     * Check availability excluding a specific reservation (for updates)
     */
    public AvailabilityResponse checkAvailabilityExcluding(
            AvailabilityRequest request,
            Long excludeReservationId
    ) {
        LocalDateTime requestedTime = request.getDateReservation();
        Integer requestedSeats = request.getNombrePersonnes();

        LocalDateTime startWindow = requestedTime.minusHours(1);
        LocalDateTime endWindow = requestedTime.plusHours(1);

        List<Reservation> reservations = reservationRepository
                .findByDateReservationBetween(startWindow, endWindow);

        // Calculate reserved seats excluding the reservation being updated
        int reservedSeats = reservations.stream()
                .filter(r -> r.getStatut() != StatutReservation.ANNULEE)
                .filter(r -> !r.getId().equals(excludeReservationId))
                .mapToInt(Reservation::getNombrePersonnes)
                .sum();

        int availableSeats = restaurantCapacity - reservedSeats;
        boolean isAvailable = availableSeats >= requestedSeats;

        return new AvailabilityResponse(
                requestedTime,
                restaurantCapacity,
                reservedSeats,
                availableSeats,
                isAvailable
        );
    }
}
